lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,42 +38,88 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not (isinstance(text, str) and isinstance(end_of_word, str)) or not (
+            isinstance(start_of_word, str) or start_of_word is None
+    ):
+        return None
+    freq_dict = {}
+    words = text.split()
+    for word in words:
+        preprocessed_word = prepare_word(word, start_of_word, end_of_word)
+        if not preprocessed_word:
+            return None
+        freq_dict.update({preprocessed_word: words.count(word)})
+    return freq_dict
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pairs_of_tokens = {}
+    for word, freq in word_frequencies.items():
+        for ind in range(len(word) - 1):
+            token_pair = (word[ind], word[ind + 1])
+            if token_pair not in pairs_of_tokens:
+                pairs_of_tokens[token_pair] = freq
```

Comment:
Is this the correct value for initizaliting the token key in the dictionary?

Reply:
fixed...